text/0001-remove_ambiguous_grammar.md

```diff
@@ -0,0 +1,234 @@
+
+# 3. Proposal
+
+## Location specific reserved words.
+
+The first change proposed is to disallow raw identifiers in places where the same identifier could also be parsed as a keyword. For example, the following code would no longer be allowed:
+
+```sql
+use = 32;
+```
+
+`use` in this case could be the start of a USE statement in this position and would therefore not be allowed. In order to get the same code one would need to surround the identifier.
+
+```sql
+`use` = 32;
+```
+
+Another example would be:
+
+```sql
+SELECT field FROM use;
+```
+
+The part after `FROM` can also have statements, however the USE statement is disallowed in subqueries, therefore allowing `use` would not be a problem.
+
+This change would completely remove existing ambiguity wherever an expression could possibly be parsed as a statement. It would simplify the parser a great deal. This changes also does not require that raw identifiers are completely distinct.
+
+## Disallow raw identifiers to start with a digit
+
+This is a common way to distinguish between a digit and number. By disallowing the first character from being a digit we can be sure that Identifiers aren’t numbers. This would resolve the same text from being used as an identifier or a number depending on the place in the grammar.
+
+## Introduce strand prefixes
+
+The only way can currently be distinguished is if one strand fails to parse as any of the other types. This can lead to problems where a use wanted a certain value to be a plain strand but it happened to match another type of strand and is thus converted into that specific type.
+
+And example is `"5:00"` which brought up in an issue. The user wanted to store this value as a plain strand but it happened to match a thing strand.
+
+I propose we introduce specific strand prefixes for specific strand types:
+
+- `t"2012-05-26T12-00-00Z"` for date-time strands
+- `u'7c3f4ce8-83c4-458f-b1d7-b28352dea93c'` for UUID’s
+- `r"5:00"` for record id strings
+
+## Syntax Error for Block-RecordId Object ambiguity
+
+The following statement is ambiguous:
+
+```sql
+{ a:b.c }
+```
+
+This can be parsed either as an object with a field `a` and value `b.c`, or as a block statement taking the field `c` of the record `a:b`. Removing this ambiguity completely would require making significant changes to how record IDs are joined (changing the `:` to something else) or requiring that either blocks or objects always be enclosed by parentheses. Both of these changes are quite drastic for a relatively minor ambiguity.
+
+Therefore, I propose that if the parser encounters `{` `Identifier` `:`, it raises a syntax error to notify the user of the ambiguity. The user would then need to either use a record ID strand if they intended to create a block, or make the identifier a strand if they intended to create an object.
```

Comment:
To visualize, would it be like:
```sql
{ r"some_record_id":b.c  }
```
or
```sql
{ a: u"some_uuid" }
```

Reply:
This ambiguity is unrelated to the ambiguity of strands.

This is a ambiguity is caused by the fact that raw records are allowed. you can form example write `CREATE some_table:some_specific_id;`. The problem is the `:`. If a raw record is the first statement in an block statement it looks exactly like an object. 

In the example `{ a:b.c }` `a:b` looks like a raw record but it could also be a field definition with a value. 

Raw records will still be allowed in these proposed changes, so we need a way to define the value of `{ a:b.c }`. Currently the parse will first try to match an object and then try to match a block statement. This results in strange behavior like `{a:b.c; true }` being parsed as an block but `{a:b.c }` is parsed as an object.

The proposed change here is to define that if the parsers sees `{` followed by and identifier followed by `:` it will always choose to parse an object and error if it encounters `{a:b.c; true }`. To make this statement work again you could use a record strand: `{ r"a:b".c; true }`.

Reply:
Ok, I think I see now. So the last code snippet at the end of your comment was what I was trying to understand. 

So the 2 options you mentioned in the rfc would be visualized as `{ r"a:b".c; true }` or  `{ a:r"b.c"; true }`?